### ISSUE TITLE
Error When Using Accelerate

### ISSUE THREAD

**Issue by ycformal** [lucidrains/denoising-diffusion-pytorch#345]
I used accelerate launch train.py.
```
/home/hice1/yxu846/.conda/envs/dog/lib/python3.9/site-packages/accelerate/accelerator.py:457: FutureWarning: Passing the following arguments to `Accelerator` is deprecated and will be removed in version 1.0 of Accelerate: dict_keys(['split_batches']). Please pass an `accelerate.DataLoaderConfiguration` instead:
dataloader_config = DataLoaderConfiguration(split_batches=True)
  warnings.warn(
/home/hice1/yxu846/.conda/envs/dog/lib/python3.9/site-packages/accelerate/accelerator.py:457: FutureWarning: Passing the following arguments to `Accelerator` is deprecated and will be removed in version 1.0 of Accelerate: dict_keys(['split_batches']). Please pass an `accelerate.DataLoaderConfiguration` instead:
dataloader_config = DataLoaderConfiguration(split_batches=True)
  warnings.warn(
atl1-1-03-012-28-0:191739:191739 [0] NCCL INFO Bootstrap : Using ib0:172.27.70.156<0>
atl1-1-03-012-28-0:191739:191739 [0] NCCL INFO NET/Plugin : Plugin load (libnccl-net.so) returned 2 : libnccl-net.so: cannot open shared object file: No such file or directory
atl1-1-03-012-28-0:191739:191739 [0] NCCL INFO NET/Plugin : No plugin found, using internal implementation
atl1-1-03-012-28-0:191739:191739 [0] NCCL INFO cudaDriverVersion 12040
NCCL version 2.18.1+cuda12.1
atl1-1-03-012-28-0:191740:191740 [0] NCCL INFO cudaDriverVersion 12040
atl1-1-03-012-28-0:191740:191740 [0] NCCL INFO Bootstrap : Using ib0:172.27.70.156<0>
atl1-1-03-012-28-0:191740:191740 [0] NCCL INFO NET/Plugin : Plugin load (libnccl-net.so) returned 2 : libnccl-net.so: cannot open shared object file: No such file or directory
atl1-1-03-012-28-0:191740:191740 [0] NCCL INFO NET/Plugin : No plugin found, using internal implementation
atl1-1-03-012-28-0:191739:191783 [0] NCCL INFO NET/IB : Using [0]mlx5_0:1/IB [RO]; OOB ib0:172.27.70.156<0>
atl1-1-03-012-28-0:191739:191783 [0] NCCL INFO Using network IB
atl1-1-03-012-28-0:191740:191784 [0] NCCL INFO NET/IB : Using [0]mlx5_0:1/IB [RO]; OOB ib0:172.27.70.156<0>
atl1-1-03-012-28-0:191740:191784 [0] NCCL INFO Using network IB

atl1-1-03-012-28-0:191740:191784 [0] init.cc:748 NCCL WARN Duplicate GPU detected : rank 1 and rank 0 both on CUDA device a000

atl1-1-03-012-28-0:191739:191783 [0] init.cc:748 NCCL WARN Duplicate GPU detected : rank 0 and rank 1 both on CUDA device a000
atl1-1-03-012-28-0:191740:191784 [0] NCCL INFO init.cc:1309 -> 5
atl1-1-03-012-28-0:191739:191783 [0] NCCL INFO init.cc:1309 -> 5
atl1-1-03-012-28-0:191740:191784 [0] NCCL INFO group.cc:64 -> 5 [Async thread]
atl1-1-03-012-28-0:191739:191783 [0] NCCL INFO group.cc:64 -> 5 [Async thread]
atl1-1-03-012-28-0:191740:191740 [0] NCCL INFO group.cc:422 -> 5
atl1-1-03-012-28-0:191739:191739 [0] NCCL INFO group.cc:422 -> 5
atl1-1-03-012-28-0:191740:191740 [0] NCCL INFO group.cc:106 -> 5
atl1-1-03-012-28-0:191739:191739 [0] NCCL INFO group.cc:106 -> 5
Traceback (most recent call last):
Traceback (most recent call last):
  File "/storage/ice1/6/1/yxu846/diffusion/denoising-diffusion-pytorch/train.py", line 16, in <module>
  File "/storage/ice1/6/1/yxu846/diffusion/denoising-diffusion-pytorch/train.py", line 16, in <module>
    trainer = Trainer(
  File "/storage/ice1/6/1/yxu846/diffusion/denoising-diffusion-pytorch/denoising_diffusion_pytorch/denoising_diffusion_pytorch.py", line 971, in __init__
    trainer = Trainer(
  File "/storage/ice1/6/1/yxu846/diffusion/denoising-diffusion-pytorch/denoising_diffusion_pytorch/denoising_diffusion_pytorch.py", line 971, in __init__
    self.model, self.opt = self.accelerator.prepare(self.model, self.opt)
  File "/home/hice1/yxu846/.conda/envs/dog/lib/python3.9/site-packages/accelerate/accelerator.py", line 1326, in prepare
    self.model, self.opt = self.accelerator.prepare(self.model, self.opt)
  File "/home/hice1/yxu846/.conda/envs/dog/lib/python3.9/site-packages/accelerate/accelerator.py", line 1326, in prepare
    result = tuple(
  File "/home/hice1/yxu846/.conda/envs/dog/lib/python3.9/site-packages/accelerate/accelerator.py", line 1327, in <genexpr>
    result = tuple(
  File "/home/hice1/yxu846/.conda/envs/dog/lib/python3.9/site-packages/accelerate/accelerator.py", line 1327, in <genexpr>
    self._prepare_one(obj, first_pass=True, device_placement=d) for obj, d in zip(args, device_placement)
  File "/home/hice1/yxu846/.conda/envs/dog/lib/python3.9/site-packages/accelerate/accelerator.py", line 1200, in _prepare_one
    self._prepare_one(obj, first_pass=True, device_placement=d) for obj, d in zip(args, device_placement)
  File "/home/hice1/yxu846/.conda/envs/dog/lib/python3.9/site-packages/accelerate/accelerator.py", line 1200, in _prepare_one
    return self.prepare_model(obj, device_placement=device_placement)
  File "/home/hice1/yxu846/.conda/envs/dog/lib/python3.9/site-packages/accelerate/accelerator.py", line 1450, in prepare_model
    return self.prepare_model(obj, device_placement=device_placement)
  File "/home/hice1/yxu846/.conda/envs/dog/lib/python3.9/site-packages/accelerate/accelerator.py", line 1450, in prepare_model
    model = torch.nn.parallel.DistributedDataParallel(
  File "/home/hice1/yxu846/.conda/envs/dog/lib/python3.9/site-packages/torch/nn/parallel/distributed.py", line 795, in __init__
    model = torch.nn.parallel.DistributedDataParallel(
  File "/home/hice1/yxu846/.conda/envs/dog/lib/python3.9/site-packages/torch/nn/parallel/distributed.py", line 795, in __init__
    _verify_param_shape_across_processes(self.process_group, parameters)
  File "/home/hice1/yxu846/.conda/envs/dog/lib/python3.9/site-packages/torch/distributed/utils.py", line 265, in _verify_param_shape_across_processes
        _verify_param_shape_across_processes(self.process_group, parameters)return dist._verify_params_across_processes(process_group, tensors, logger)

  File "/home/hice1/yxu846/.conda/envs/dog/lib/python3.9/site-packages/torch/distributed/utils.py", line 265, in _verify_param_shape_across_processes
torch.distributed.DistBackendError: NCCL error in: ../torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1333, invalid usage (run with NCCL_DEBUG=WARN for details), NCCL version 2.18.1
ncclInvalidUsage: This usually reflects invalid usage of NCCL library.
Last error:
Duplicate GPU detected : rank 0 and rank 1 both on CUDA device a000
    return dist._verify_params_across_processes(process_group, tensors, logger)
torch.distributed.DistBackendError: NCCL error in: ../torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1333, invalid usage (run with NCCL_DEBUG=WARN for details), NCCL version 2.18.1
ncclInvalidUsage: This usually reflects invalid usage of NCCL library.
Last error:
Duplicate GPU detected : rank 1 and rank 0 both on CUDA device a000
[2024-10-07 11:29:17,757] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 191739 closing signal SIGTERM
[2024-10-07 11:29:17,871] torch.distributed.elastic.multiprocessing.api: [ERROR] failed (exitcode: 1) local_rank: 1 (pid: 191740) of binary: /home/hice1/yxu846/.conda/envs/dog/bin/python
Traceback (most recent call last):
  File "/home/hice1/yxu846/.conda/envs/dog/bin/accelerate", line 8, in <module>
    sys.exit(main())
  File "/home/hice1/yxu846/.conda/envs/dog/lib/python3.9/site-packages/accelerate/commands/accelerate_cli.py", line 48, in main
    args.func(args)
  File "/home/hice1/yxu846/.conda/envs/dog/lib/python3.9/site-packages/accelerate/commands/launch.py", line 1165, in launch_command
    multi_gpu_launcher(args)
  File "/home/hice1/yxu846/.conda/envs/dog/lib/python3.9/site-packages/accelerate/commands/launch.py", line 799, in multi_gpu_launcher
    distrib_run.run(args)
  File "/home/hice1/yxu846/.conda/envs/dog/lib/python3.9/site-packages/torch/distributed/run.py", line 797, in run
    elastic_launch(
  File "/home/hice1/yxu846/.conda/envs/dog/lib/python3.9/site-packages/torch/distributed/launcher/api.py", line 134, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/home/hice1/yxu846/.conda/envs/dog/lib/python3.9/site-packages/torch/distributed/launcher/api.py", line 264, in launch_agent
    raise ChildFailedError(
torch.distributed.elastic.multiprocessing.errors.ChildFailedError:
============================================================
train.py FAILED
------------------------------------------------------------
Failures:
  <NO_OTHER_FAILURES>
------------------------------------------------------------
Root Cause (first observed failure):
[0]:
  time      : 2024-10-07_11:29:17
  host      : atl1-1-03-012-28-0.pace.gatech.edu
  rank      : 1 (local_rank: 1)
  exitcode  : 1 (pid: 191740)
  error_file: <N/A>
  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html
============================================================
```

**Comment by lucidrains** [comment]
@ycformal could you try downgrading accelerate to a version that works? let me know which one and i'll just pin the version for now

**Comment by ycformal** [comment]
The problem here is I only applied one GPU on my cluster. After applying two, it works. However, before that, I downgraded the torch version to 2.1.2+cu121, installed nccl, and disabled all optimizations on accelerate config. I don't know which of them work. Using torch 2.4 results an error when saving the model. The nccl python package is automatically installed when running `pip install torch`.

In summary, the following operations work for me:
Install torch 2.1.2
Install the nvidia-nccl package via pip and install nccl on the nvidia website
Ensure there are multiple GPUs
Disable all optimizations when configuring accelerate

Some ablation studies are needed to get the necessary environment condition.

**Comment by lucidrains** [comment]
@ycformal what was the warning for 2.4?

**Comment by ycformal** [comment]
> @ycformal what was the warning / error for 2.4?

It's not recorded.. I just heard somewhere that nccl has some problems with torch 2.4 and followed the advice to downgrade torch to 2.1.2.

**Comment by lucidrains** [comment]
ah got it
